### DISCUSSION TITLE
USB hub device?

### DISCUSSION THREAD

**Discussion by rhulme** [hathach/tinyusb#1679] [Q&A] [answered]
Is it possible or feasible to implement a USB hub as a device?

This would allow, for example, the uC to implement its own device interfaces whilst also making other devices available attached to a second USB port (working in host mode).

Alternatively, a uC could implement an existing protocol associated with a specific VID/PID (for compatibility with existing software/drivers) while also adding additional functionality of its own under a different VID/PID pair.

**Comment by hathach** [comment, marked as answer]
I don't think it is possible since hub works mostly in the electrical layer. But you could try your luck to implement an limited hub.

**Reply by rhulme** [reply]
Thank you for the quick response.

My primary interest is actually in the second example I gave (one uC implementing devices under multiple VID/PID pairs).

I only have a passing knowledge of the details of how USB actually works but my understanding was that a hub must also appear as a device of some form as it has to present a control channel and the host systems knows that the devices are behind the hub, so it's therefore not completely transparent.  It seems it would be "base class 09h".

Assuming that device class were implemented, there would still be issue of how to present two (or theoretically more) separate devices.  Is that something that is even feasible with the current architecture?  I guess it would be like having two instances of the stack running but sharing a single hardware interface?

Apologies if I'm getting concepts completely mixed up.  If there are any specific keywords I can google, or files in tinyusb I should look at, I'm happy to go digging.

**Reply by rhulme** [reply]
Having hunted around a bit, it seems that the keyword is a "compound device", which is effectively just a hub with permanently attached devices.

If I understand the sequence correctly, the hub allows the host to individually reset connected devices, which sets their device address to 0.  The host then speaks directly to device 0 (of which there should be only one) and assigns it a new address, which is then used for all further communication.

So the primary requirement for this to work would be for the USB hardware module on the uC to support more than one device address otherwise it will never respond.  Even if that were possible (and it doesn't look like it), it's probably not easy to add that concept to tinyusb.  Functions like `tud_descriptor_configuration_cb()` would need to include the device address or similar so that the correct structure could be returned.

Pity...

**Reply by hathach** [reply]
yeah for multiple devices/phys, I answered something similar here #894
